proposals/0162-remove-accounts-executable-flag-checks.md

```diff
@@ -0,0 +1,137 @@
+
+Removing the first condition not only removes some redundant error checks from
+the runtime and confusion for future maintainers, but more importantly enables
+us to deploy loader-v4 which is upgradeable like loader-v3 but without the need
+for proxy accounts.
+
+## New Terminology
+
+None.
+
+## Detailed Design
+
+This proposal aims to unblock loader-v4 with minimal impact on the ecosystem by
+only removing checks of the `is_executable` flag. A complete removal of the
+flag can be addressed in a subsequent proposal. Thus, the following must remain
+unaffected for now:
+
+- Setting of the `is_executable` flag during program deployment
+- Calculation of the account hashes
+- Minimization of snapshots
+- Serialization of instruction accounts `is_executable` flag for dapps
+- CPI ignores changes made by the caller to instruction accounts which have
+the `is_executable` flag set
+
+These checks of the `is_executable` flag must be removed:
+
+- `ExecutableLamportChange` during execution (transaction succeeds instead)
+
+These checks of the `is_executable` flag are irrelevant to consensus as
+transactions which try to invoke accounts, which do not contain programs,
+continue to fail without them. They should still be removed because all
+implementations should aim to produce the same error codes:
+
+- during transaction loading:
+  - `InvalidProgramForExecution` (fallthrough to `UnsupportedProgramId` during
+  execution)
+  - `AccountNotExecutable` (fallthrough to `UnsupportedProgramId` during
+  execution)
+  - Meaning the transaction loading checks are effectively deferred until
+execution, which gives users more complete transaction logs.
+
+- in the `Upgrade` instruction of loader-v3:
+  - `AccountNotExecutable`
+  (fallthrough to `IncorrectProgramId` or `InvalidAccountData`, depending on
+  the owner)
+
+- during execution:
+  - `AccountNotExecutable` (fallthrough to `UnsupportedProgramId` during
+  execution)
+  - `IncorrectProgramId` (fallthrough to `UnsupportedProgramId` during
+  execution)
+  - `ExecutableDataModified`
+  (fallthrough to `ExternalAccountDataModified` during execution)
+  - `ExecutableModified`
+  (is unreachable, as it is overshadowed by owner and writability checks)
+  - `ModifiedProgramId`
+  (is unreachable, as it is overshadowed by owner and writability checks)
+
+Similarly the following checks during execution, unrelated to the
+`is_executable` flag, but related to whether an account contains a program or
+not, should be changed to throw `UnsupportedProgramId` instead:
+
+- `InvalidAccountData` for programs which are closed, in visibility delay,
+failed verification or not owned by a loader
+- `IncorrectProgramId` for unrecognized built-in programs
+
+All in all, the following error messages related to invocation of program
+accounts will be coalesced into `UnsupportedProgramId`:
+
+- `InvalidProgramForExecution`
+- `AccountNotExecutable`
+- `InvalidAccountData`
+- `IncorrectProgramId`
+- `UnsupportedProgramId`
```

Comment:
Will these be deprecated, then?

Reply:
Some of them (such as `InvalidAccountData`) can still be produced by other built-in / core programs, e.g. program management instructions.

Reply:
Okay sounds good. No change needed here imo, then.